src/Microsoft.ML.Data/Data/SchemaDefinition.cs

```diff
@@ -389,9 +402,14 @@ public static SchemaDefinition Create(Type userType, Direction direction = Direc
                 var vectorAttr = memberInfo.GetCustomAttribute<VectorTypeAttribute>();
                 if (vectorAttr != null && !isVector)
                     throw Contracts.ExceptParam(nameof(userType), "Member {0} marked with VectorType attribute, but does not appear to be a vector type", memberInfo.Name);
+                var varVectorAttr = memberInfo.GetCustomAttribute<VariableVectorTypeAttribute>();
+                if (varVectorAttr != null && !isVector)
+                    throw Contracts.ExceptParam(nameof(userType), "Member {0} marked with VariableVectorType attribute, but does not appear to be a vector type", memberInfo.Name);
+                if (vectorAttr != null && varVectorAttr != null)
+                    throw Contracts.ExceptParam(nameof(userType), "Member {0} marked with both VariableVectorType and VectorType attribute", memberInfo.Name);
                 if (isVector)
                 {
-                    int[] dims = vectorAttr?.Dims;
+                    int[] dims = varVectorAttr != null ? new int[0] : vectorAttr?.Dims;
```

Comment:
>vectorAttr?.Dims; [](start = 70, length = 17)

This one makes me question do we even needed VariableVectorType attribute.
In past, and now we just treat any array as variable one, and presence of VariableVectorType is a bit wacky.
I can probably force system to fail if it's a vector and none of attributes is present.
Or I can just specify in documentation for VectorType what absence of VectorType means this array has unfixed size.

Reply:
Maybe some people get nervous without explicit attributes identifying types, I don't know. I am unable to trace back its addition or any discussion regarding it. If we wanted an explicit attribute for *some* reason, maybe we can just get away with vector type attribute with no arguments? Which I think is also how our vector type actually does behave.